ui/hooks/gator-permissions/useGatorPermissionTokenInfo.ts

```diff
@@ -0,0 +1,270 @@
+import { getAllNetworkConfigurationsByCaipChainId } from '../../../shared/modules/selectors/networks';
+import { getTokenStandardAndDetailsByChain } from '../../store/actions';
+import {
+  resolveNativeTokenInfo,
+  lookupCachedOrImportedTokenInfo,
+  getGatorErc20TokenInfo,
+  type GatorTokenInfo,
+  type CachedTokensByChain,
+  type ImportedTokensByChain,
+  type NetworkConfigurationsByCaipChainId,
+} from '../../../shared/lib/gator-permissions/gator-permissions-utils';
+
+export type { GatorTokenInfo } from '../../../shared/lib/gator-permissions/gator-permissions-utils';
+
+type TokenInfoSource =
+  | 'native'
+  | 'cache'
+  | 'imported'
+  | 'api'
+  | 'onchain'
+  | null;
+
+export type UseGatorPermissionTokenInfoResult = {
+  /**
+   * Token information including symbol, decimals, name, and image
+   */
+  tokenInfo: GatorTokenInfo;
+  /**
+   * True when fetching token information
+   */
+  loading: boolean;
+  /**
+   * Error that occurred during fetch, if any
+   */
+  error: Error | null;
+  /**
+   * The source from which token info was retrieved
+   * - 'native': Native token from network configuration
+   * - 'cache': From tokensChainsCache (API cache)
+   * - 'imported': From user's imported tokens
+   * - 'api': From external API
+   * - 'onchain': From on-chain data
+   */
+  source: TokenInfoSource;
+};
+
+/**
+ * Hook to fetch token information with a multi-tier fallback strategy:
+ * 1. Handle native tokens if permissionType contains 'native-token'
+ * 2. Check cache (tokensChainsCache) and user imported tokens
+ * 3. Fall back to API call (if external services enabled)
+ * 4. Fall back to on-chain data
+ *
+ * @param tokenAddress - The token contract address (undefined for native tokens)
+ * @param chainId - The chain ID in hex format
+ * @param permissionType - Optional permission type to detect native tokens
+ * @returns Token information with loading and error states
+ */
+export function useGatorPermissionTokenInfo(
+  tokenAddress: string | undefined,
+  chainId: Hex | undefined,
+  permissionType?: string,
+): UseGatorPermissionTokenInfoResult {
+  const [error, setError] = useState<Error | null>(null);
+  const [fetchedTokenInfo, setFetchedTokenInfo] = useState<{
+    tokenInfo: GatorTokenInfo;
+    source: 'api' | 'onchain';
+  } | null>(null);
+
+  // Get cached tokens from tokensChainsCache (API cache)
+  const erc20TokensByChain = useSelector(selectERC20TokensByChain);
+
+  // Get user's imported tokens
+  const allTokens = useSelector(getAllTokens);
+
+  // Get external services preference
+  const allowExternalServices = useSelector(getUseExternalServices);
+
+  // Get network configurations for native token handling
+  const networkConfigurationsByCaipChainId = useSelector(
+    getAllNetworkConfigurationsByCaipChainId,
+  );
+
+  // Check if this is a native token
+  const isNativeToken = useMemo(() => {
+    return permissionType?.includes('native-token') ?? false;
+  }, [permissionType]);
```

Comment:
nit: I'm not convinced that this needs a memo, as it's not an expensive function. Can just be calculated on each render.